lib/streamlit/connections/sql_connection.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+from datetime import timedelta
+from typing import TYPE_CHECKING, Iterator, Optional, Union, cast
+
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine.base import Engine
+    from sqlalchemy.orm import Session
+
+
+_REQUIRED_CONNECTION_PARAMS = {"dialect", "username", "host"}
+
+
+class SQL(BaseConnection["Engine"]):
+    _default_connection_name = "sql"
+
+    def connect(self, autocommit: bool = False, **kwargs) -> "Engine":
+        import sqlalchemy
+
+        secrets = self.get_secrets()
+
+        if "url" in secrets:
+            url = sqlalchemy.engine.make_url(secrets["url"])
+        else:
+            for p in _REQUIRED_CONNECTION_PARAMS:
+                if p not in secrets:
+                    raise StreamlitAPIException(f"Missing SQL DB connection param: {p}")
+
+            drivername = secrets["dialect"] + (
+                f"+{secrets['driver']}" if "driver" in secrets else ""
+            )
+
+            url = sqlalchemy.engine.URL.create(
+                drivername=drivername,
+                username=secrets["username"],
+                password=secrets.get("password"),
+                host=secrets["host"],
+                port=int(secrets["port"]) if "port" in secrets else None,
+                database=secrets.get("database"),
+            )
+
+        eng = sqlalchemy.create_engine(url, **kwargs)
```

Comment:
I still believe we need some way for additional secrets to be added as arguments in this call.

A good concrete example is BigQuery. I'd like to be able to configure my `credentials_info` in secrets.toml or provide a `credentials_path` and have them passed to `create_engine()`.

Cite: https://github.com/googleapis/python-bigquery-sqlalchemy#authentication

Reply:
Ack -- I'll address this when circling back to merging config options / params from various places for both this connector and the Snowpark one